### PR TITLE
Chainer graph

### PR DESCRIPTION
**Pull request by knshnb**
## New features
- New data(set) patterns
  - sparse graph data(set)
  - padding graph data(set) with coo matrix
- New models
  - sparse RelGCN
  - sparse GIN
  - GIN with coo matrix
- New task
  - semi-supervised node classification on large graphs
- New datasets
  - citation network (cora, citeseer)
  - reddit

## Dataset class
| | Chemical | Network
----|----|----
| adjacency matrix | `NumpyTupleDataset` | `PaddingGraphDataset` | 
| scatter operation | `SparseGraphDataset` | `SparseGraphDataset` | 
| sparse matmul | not supported | `PaddingGraphDataset (use_coo=True)` |

### CODE REVIEW

File: chainer_chemistry/dataset/preprocessors/ggnn_preprocessor.py [hunk 1]
```diff
@@ -45,4 +51,45 @@ def get_input_features(self, mol):
         type_check_num_atoms(mol, self.max_atoms)
         atom_array = construct_atomic_number_array(mol, out_size=self.out_size)
         adj_array = construct_discrete_edge_matrix(mol, out_size=self.out_size)
+
         return atom_array, adj_array
+
+    def create_dataset(self, *args, **kwargs):
```
Comment (corochann) [inline chainer_chemistry/dataset/preprocessors/ggnn_preprocessor.py]: is it possible to return NumpyTupleDataset for padding pattern dataset for now?

File: chainer_chemistry/dataset/preprocessors/mol_preprocessor.py [hunk 2]
```diff
@@ -93,6 +94,9 @@ def get_input_features(self, mol):
         """
         raise NotImplementedError
 
+    def create_dataset(self, *args, **kwargs):
+        return NumpyTupleDataset(*args)
```
Comment (corochann) [inline chainer_chemistry/dataset/preprocessors/mol_preprocessor.py]: LGTM!!

File: chainer_chemistry/datasets/citation_network/citation.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    G = nx.convert_node_labels_to_integers(G)
+    print("Finished loading graph: {}".format(filepath))
+    print("number of nodes: {}, number of edges: {}".format(
+        G.number_of_nodes(), G.number_of_edges()
+    ))
+    return G
+
+
+def cora_to_networkx():
```
Comment (corochann) [inline chainer_chemistry/datasets/citation_network/citation.py]: maybe specify filepath from args

File: chainer_chemistry/datasets/citation_network/citation.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    return _to_networkx(
+        "chainer_chemistry/datasets/citation_network/cora/",
+        "cora"
+    )
+
+
+def citeseer_to_network():
```
Comment (corochann) [inline chainer_chemistry/datasets/citation_network/citation.py]: maybe specify filepath from args

File: chainer_chemistry/datasets/citation_network/citeseer/.gitignore [hunk 5]
```diff
@@ -0,0 +1,3 @@
+citeseer.cites
```
Comment (corochann) [inline chainer_chemistry/datasets/citation_network/citeseer/.gitignore]: remove this file. I think dataset should be inside `examples` directory, not inside library.

File: chainer_chemistry/datasets/citation_network/cora/.gitignore [hunk 6]
```diff
@@ -0,0 +1,3 @@
+cora.cites
```
Comment (corochann) [inline chainer_chemistry/datasets/citation_network/cora/.gitignore]: remove this file. I think dataset should be inside `examples` directory, not inside library.

File: chainer_chemistry/datasets/numpy_tuple_dataset.py [hunk 7]
```diff
@@ -48,6 +49,10 @@ def __len__(self):
     def get_datasets(self):
         return self._datasets
 
+    @property
+    def converter(self):
+        return concat_mols
```
Comment (corochann) [inline chainer_chemistry/datasets/numpy_tuple_dataset.py]: TODO later: consider design. NumpyTupleDataset may be used with other converter.

File: chainer_chemistry/datasets/reddit/.gitignore [hunk 8]
```diff
@@ -0,0 +1,3 @@
+reddit.zip
```
Comment (corochann) [inline chainer_chemistry/datasets/reddit/.gitignore]: remove this file. I think dataset should be inside `examples` directory, not inside library.

File: chainer_chemistry/datasets/reddit/reddit.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+import scipy
+
+
+def _to_networkx(filepath):
+    print("Loading graph data")
+    coo_adj = scipy.sparse.load_npz(filepath + "reddit_graph.npz")
+    G = nx.from_scipy_sparse_matrix(coo_adj)
+
+    print("Loading node feature and label")
+    # 頂点feature, 頂点label
```
Comment (corochann) [inline chainer_chemistry/datasets/reddit/reddit.py]: can you change to English?

File: chainer_chemistry/links/connection/mlp.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
+import chainer
```
Comment (corochann) [inline chainer_chemistry/links/connection/mlp.py]: I think MLP is defined inside `models`

File: chainer_chemistry/links/__init__.py [hunk 11]
```diff
@@ -1,6 +1,7 @@
 from chainer_chemistry.links.connection.embed_atom_id import EmbedAtomID  # NOQA
 from chainer_chemistry.links.connection.graph_linear import GraphLinear  # NOQA
 from chainer_chemistry.links.connection.graph_mlp import GraphMLP  # NOQA
+from chainer_chemistry.links.connection.mlp import MLP  # NOQA
```
Comment (corochann) [inline chainer_chemistry/links/__init__.py]: MLP is defined in models, remove this.

File: chainer_chemistry/links/update/gin_update.py [hunk 12]
```diff
@@ -3,7 +3,7 @@
 
 import chainer_chemistry
 from chainer_chemistry.links import GraphMLP
-from chainer_chemistry.links.connection.graph_linear import GraphLinear
+from chainer_chemistry.links import MLP
```
Comment (corochann) [inline chainer_chemistry/links/update/gin_update.py]: use models.MLP

File: chainer_chemistry/models/ggnn.py [hunk 13]
```diff
@@ -55,7 +55,7 @@ def __init__(self, out_dim, hidden_channels=16, n_update_layers=4,
         self.concat_hidden = concat_hidden
         self.weight_tying = weight_tying
 
-    def __call__(self, atom_array, adj, is_real_node=None):
+    def __call__(self, batch):
```
Comment (corochann) [inline chainer_chemistry/models/ggnn.py]: can you keep backward compatibility?

File: chainer_chemistry/models/gin.py [hunk 14]
```diff
@@ -28,10 +29,12 @@ class GIN(chainer.Chain):
...
             Defaults to 4 for single, double, triple and aromatic bond.
     """
-    def __init__(self, out_dim, hidden_channels=16,
+
+    def __init__(self, out_dim, node_embedding=False, hidden_channels=16,
+                 out_channels=None,
                  n_update_layers=4, n_atom_types=MAX_ATOMIC_NUM,
                  dropout_ratio=0.5, concat_hidden=False,
-                 weight_tying=True, activation=functions.identity,
+                 weight_tying=False, activation=functions.identity,
```
Comment (corochann) [inline chainer_chemistry/models/gin.py]: not change default value

File: chainer_chemistry/models/relgcn.py [hunk 15]
```diff
@@ -79,14 +82,16 @@ def __init__(self, out_dim=64, hidden_channels=None, n_update_layers=None,
         self.input_type = input_type
         self.scale_adj = scale_adj
 
-    def __call__(self, x, adj):
+    def __call__(self, batch):
```
Comment (corochann) [inline chainer_chemistry/models/relgcn.py]: can we keep backward compatibility?

Comment (corochann) [reply chainer_chemistry/models/relgcn.py]: use NumpyTupleDataset for dense data in chemistry dataset, and use wrapper Chain for dense SNS dataset example??

File: chainer_chemistry/models/prediction/node_classifier.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+    """A simple node classifier model.
+    """
+
+    def __call__(self, data, train_mask, valid_mask, *args, **kwargs):
+        """Computes the loss value for an input and label pair.
+        """
+        self.metrics = None
+        if isinstance(data, PaddingGraphData):
+            self.y = self.predictor(data.x, data.adj)
```
Comment (corochann) [inline chainer_chemistry/models/prediction/node_classifier.py]: maybe okay to use `self.predictor(data)` and wrap Chain side in SNS dataset example.

File: chainer_chemistry/utils/train_utils.py [hunk 17]
```diff
@@ -82,3 +83,52 @@ def run_train(model, train, valid=None,
...
+def run_node_classification_train(model, data,
+                                  train_mask, valid_mask,
+                                  epoch=10,
+                                  optimizer=None,
+                                  out='result',
+                                  extensions_list=None,
+                                  device=-1,
+                                  converter=None,
+                                  use_default_extensions=True,
+                                  resume_path=None):
```
Comment (corochann) [inline chainer_chemistry/utils/train_utils.py]: MEMO. TODO (nakago) later: check  again. can't we merge to run_train?

File: examples/network_graph/README.md [hunk 18]
```diff
@@ -0,0 +1,36 @@
+# Network Node Classification Example
```
Comment (corochann) [inline examples/network_graph/README.md]: LGTM!!

Comment (corochann) [reply examples/network_graph/README.md]: If you can cite each dataset, please cite.

File: examples/qm9/train_qm9.py [hunk 19]
```diff
@@ -117,13 +116,17 @@ def main():
         # Cache the laded dataset.
         if not os.path.exists(cache_dir):
             os.makedirs(cache_dir)
-        NumpyTupleDataset.save(dataset_cache_path, dataset)
+        # NumpyTupleDataset.save(dataset_cache_path, dataset)
```
Comment (corochann) [inline examples/qm9/train_qm9.py]: if isinstance(dataset, NumpyTupleDataset)

File: examples/network_graph/train_network_graph.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+from chainer_chemistry.dataset.networkx_preprocessors.gin_reddit import get_reddit_coo_data  # NOQA
+
+
+dataset_dict = {
+    'cora': cora_to_networkx,
+    'citeseer': citeseer_to_network,
+    'reddit': reddit_to_networkx,
+}
+method_dict = {
+    'gin': GIN,
```
Comment (corochann) [inline examples/network_graph/train_network_graph.py]: wrap GIN by another Chain

```
import chainer
class PaddingDatasetWrapper(chainer.Chain):
    def __init__(self, predictor):
        super(PaddingDatasetWrapper, self).__init__()
        with self.init_scope():
            self.predictor = predictor

    def forward(self, dataset):
        assert isinstance(dataset, PaddingDataset)
        return self.predictor(dataset.x, dataset.adj)


# use as follows:
model = GIN()
model = PaddingDatasetWrapper(model)
```

File: chainer_chemistry/dataset/preprocessors/ggnn_preprocessor.py (hunk 1 above)
Comment (knshnb) [reply chainer_chemistry/dataset/preprocessors/ggnn_preprocessor.py]: fixed!

File: .gitignore [hunk 21]
```diff
@@ -113,3 +113,7 @@ examples/own_dataset/eval_*
 *.pyproj
 
 .pytest_cache
+
+# Project cache directory
```
[the remark this answers is not in the document]
Comment (knshnb) [reply .gitignore]: fixe!

File: chainer_chemistry/datasets/citation_network/citation.py (hunk 3 above)
Comment (knshnb) [reply chainer_chemistry/datasets/citation_network/citation.py]: fixed!

File: chainer_chemistry/datasets/citation_network/citation.py (hunk 4 above)
Comment (knshnb) [reply chainer_chemistry/datasets/citation_network/citation.py]: fixed!

File: chainer_chemistry/datasets/citation_network/citeseer/.gitignore (hunk 5 above)
Comment (knshnb) [reply chainer_chemistry/datasets/citation_network/citeseer/.gitignore]: fixed!

File: chainer_chemistry/datasets/citation_network/cora/.gitignore (hunk 6 above)
Comment (knshnb) [reply chainer_chemistry/datasets/citation_network/cora/.gitignore]: fixed!

File: chainer_chemistry/datasets/reddit/.gitignore (hunk 8 above)
Comment (knshnb) [reply chainer_chemistry/datasets/reddit/.gitignore]: fixed!

File: chainer_chemistry/datasets/reddit/reddit.py (hunk 9 above)
Comment (knshnb) [reply chainer_chemistry/datasets/reddit/reddit.py]: fixed!

File: chainer_chemistry/models/ggnn.py (hunk 13 above)
Comment (knshnb) [reply chainer_chemistry/models/ggnn.py]: fixed!

File: chainer_chemistry/models/gin.py (hunk 14 above)
Comment (knshnb) [reply chainer_chemistry/models/gin.py]: weight_tying has a bug and I think it should be False by default

File: chainer_chemistry/models/relgcn.py (hunk 15 above)
Comment (knshnb) [reply chainer_chemistry/models/relgcn.py]: fixed!

File: chainer_chemistry/dataset/networkx_preprocessors/gin_reddit.py [hunk 22]
```diff
@@ -0,0 +1,40 @@
+import numpy
```
[the remark this answers is not in the document]
Comment (knshnb) [reply chainer_chemistry/dataset/networkx_preprocessors/gin_reddit.py]: fixed!

File: chainer_chemistry/links/connection/mlp.py (hunk 10 above)
Comment (knshnb) [reply chainer_chemistry/links/connection/mlp.py]: fixed!

File: chainer_chemistry/links/__init__.py (hunk 11 above)
Comment (knshnb) [reply chainer_chemistry/links/__init__.py]: fixed!

File: chainer_chemistry/links/update/gin_update.py (hunk 12 above)
Comment (knshnb) [reply chainer_chemistry/links/update/gin_update.py]: fixed!

File: examples/qm9/train_qm9.py (hunk 19 above)
Comment (knshnb) [reply examples/qm9/train_qm9.py]: fixed!

File: chainer_chemistry/models/prediction/node_classifier.py (hunk 16 above)
Comment (knshnb) [reply chainer_chemistry/models/prediction/node_classifier.py]: fixed!

File: examples/network_graph/train_network_graph.py (hunk 20 above)
Comment (knshnb) [reply examples/network_graph/train_network_graph.py]: fixed!

File: chainer_chemistry/models/gin.py (hunk 14 above)
Comment (knshnb) [reply chainer_chemistry/models/gin.py]: #399

File: chainer_chemistry/models/gin.py [hunk 23]
```diff
@@ -92,12 +102,92 @@ def __call__(self, atom_array, adj, is_real_node=None):
         for step in range(self.n_message_layers):
```
Comment (corochann) [inline chainer_chemistry/models/gin.py]: self.n_update_layers to fix other issue.

File: examples/network_graph/README.md (hunk 18 above)
Comment (knshnb) [reply examples/network_graph/README.md]: fixed!

File: chainer_chemistry/models/prediction/padding_model_wrapper.py [hunk 24]
```diff
@@ -0,0 +1,13 @@
+import chainer
```
Comment (corochann) [inline chainer_chemistry/models/prediction/padding_model_wrapper.py]: sorry could you define this class in example? this is temporal class.

**Comment by corochann** [comment]
updated comment

**Comment by corochann** [comment]
updated comment

File: chainer_chemistry/models/ggnn.py [hunk 25]
```diff
@@ -71,6 +71,7 @@ def __call__(self, atom_array, adj, is_real_node=None):
         Returns:
             ~chainer.Variable: minibatch of fingerprint
         """
+        is_real_node = None  # TODO: support this
```
[the remark this answers is not in the document]
Comment (knshnb) [reply chainer_chemistry/models/ggnn.py]: fixed!

File: chainer_chemistry/models/gin.py [hunk 26]
```diff
@@ -92,12 +102,92 @@ def __call__(self, atom_array, adj, is_real_node=None):
...
+    def __call__(self, sparse_batch, is_real_node=None):
+        if sparse_batch.x.dtype == self.xp.int32:
+            h = self.embed(sparse_batch.x)  # (minibatch, max_num_atoms)
+        else:
+            h = self.first_mlp(sparse_batch.x)
+
+        h0 = functions.copy(h, cuda.get_device_from_array(h.data).id)
+
+        g_list = []
+        for step in range(self.n_message_layers):
```
[the remark this answers is not in the document]
Comment (knshnb) [reply chainer_chemistry/models/gin.py]: fixed!

File: chainer_chemistry/models/gin.py (hunk 23 above)
Comment (knshnb) [reply chainer_chemistry/models/gin.py]: fixed!

File: chainer_chemistry/models/prediction/padding_model_wrapper.py (hunk 24 above)
Comment (knshnb) [reply chainer_chemistry/models/prediction/padding_model_wrapper.py]: fixed!

**Comment by corochann** [comment]
Thank you! LGTM. I will check to run example later!

**Comment by corochann** [comment]
If you can add some test, please add test (coverage test is failing now, since new functionality is not tested.) But I am going to merge this PR, even though in current condition after I confirmed example code works.

File: examples/network_graph/README.md [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+To train a model, run the following:
+
+On the CPU:
+```angular2html
+PYTHONPATH=. python examples/network_graph/train_network_graph.py --dataset cora
+```
+
+On the GPU:
+```angular2html
+PYTHONPATH=. python examples/network_graph/train_network_graph.py --dataset cora
```
Comment (corochann) [inline examples/network_graph/README.md]: I will add `--device 0` later.

File: examples/network_graph/README.md [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+```
```
Comment (corochann) [inline examples/network_graph/README.md]: todo later: I'd like to modify running this script from `examples/network_graph` instead of top directory.

Comment (corochann) [reply examples/network_graph/README.md]: todo later: I'd like to add `--coo` option explanation, which is necessary to run reddit dataset.

Comment (corochann) [reply examples/network_graph/README.md]: todo later: it seems `gin_sparse` and `coo=true` combination is not supported.


reddit dataset only works with --method gin and --coo true.

**Comment by corochann** [comment]
I will merge this PR!

Known TODO: #400 #401

**Pull request merged**